src/dialect.rs

```diff
@@ -0,0 +1,73 @@
+use crate::allocator::{Allocator, NodePtr};
+use crate::cost::Cost;
+use crate::operator_handler::OperatorHandler;
+use crate::reduction::Response;
+
+use crate::run_program::{run_program, PreEval};
+
+pub struct Dialect<Handler: OperatorHandler> {
+    quote_kw: Vec<u8>,
+    apply_kw: Vec<u8>,
+    op_handler: Handler,
+}
+
+impl<Handler: OperatorHandler> OperatorHandler for Dialect<Handler> {
+    fn op(
+        &self,
+        allocator: &mut Allocator,
+        op: NodePtr,
+        args: NodePtr,
+        max_cost: Cost,
+    ) -> Response {
+        self.op_handler.op(allocator, op, args, max_cost)
+    }
+}
+
+impl<Handler: OperatorHandler> Dialect<Handler> {
+    pub fn new(quote_kw: &[u8], apply_kw: &[u8], op_handler: Handler) -> Self {
+        Dialect {
+            quote_kw: quote_kw.to_owned(),
+            apply_kw: apply_kw.to_owned(),
+            op_handler,
+        }
+    }
+
+    pub fn run_program_with_pre_eval(
+        &self,
+        allocator: &mut Allocator,
+        program: NodePtr,
+        args: NodePtr,
+        max_cost: Cost,
+        pre_eval: PreEval,
+    ) -> Response {
+        run_program(
+            allocator,
+            &self.quote_kw,
+            &self.apply_kw,
+            self,
+            program,
+            args,
+            max_cost,
+            Some(pre_eval),
+        )
+    }
+
+    pub fn run_program(
```

Comment:
is there a good reason to wrap `run_program()` in this class? Why not pass in the dialect to `run_program()` directly?

Reply:
Mostly convenience... `Dialect` was always intended to be the highest level object... so most users will just go ```
let dialect = chia_dialect(true);
let r = dialect.run_program(...)
```

Reply:
```
let r = run_program(chia_dialect(true), ...);
```
is pretty convenient too. The problem with member functions is that every dialect will have to implement these forwarding functions, or derive from a class that that does (I'm not sure exactly how that works in rust, but I'm sure it's possible)

Reply:
I sketched out what it would look like making `Dialect` a `trait`, but it didn't really make things simpler, so still it's a (very simple) `struct`. So making it a type parameter to `RunProgramContext` would be a circular dependency between `dialect.rs` and `run_program.rs`, which smells bad to me.

Reply:
you get the circular dependency because you add these member functions to `Dialect`. The idea of `Dialect` being a trait is that *that's* the interface/customization point.